### PR TITLE
feat: show restricted chats and suspicious treatment updates

### PR DESCRIPTION
**Pull request by iProdigy**
Implementation of https://gist.github.com/iProdigy/fa26cf4739c872b8d23cfb861b139079

Screenshot_20231229_114034

Screenshot_20231229_200909

todo: use pubsub topic to annotate monitored chat messages (and delete the corresponding message sent over irc)
todo: show user chat badges 
todo: show emotes in caught message
todo: add search flags

(these todos are likely for separate follow-up PRs)

### CODE REVIEW

File: src/Application.cpp [hunk 1]
```diff
@@ -473,6 +474,68 @@ void Application::initPubSub()
                 });
             });
 
+    std::ignore =
+        this->twitch->pubsub->signals_.moderation.suspiciousMessageReceived
+            .connect([&](const auto &action) {
+                if (action.treatment !=
+                    PubSubLowTrustUsersMessage::Treatment::Restricted)
+                {
+                    return;
```
Comment (pajlada) [inline src/Application.cpp:483]: Maybe add a comment here stating that we currently only show restricted messages, but that we have plans to highlight monitored users soon(tm)

File: src/Application.cpp [hunk 2]
```diff
@@ -473,6 +474,68 @@ void Application::initPubSub()
...
+            });
+
+    std::ignore =
+        this->twitch->pubsub->signals_.moderation.suspiciousTreatmentUpdated
+            .connect([&](const auto &action) {
+                if (action.updatedByUserLogin.isEmpty() ||
+                    action.treatment ==
+                        PubSubLowTrustUsersMessage::Treatment::INVALID)
+                {
+                    return;
```
Comment (pajlada) [inline src/Application.cpp:516]: Maybe a qcwarn here or something?

File: src/Application.cpp [hunk 3]
```diff
@@ -473,6 +474,68 @@ void Application::initPubSub()
...
+                if (chan->isEmpty())
+                {
+                    return;
+                }
+
+                auto msg =
+                    TwitchMessageBuilder::makeLowTrustUpdateMessage(action);
+                postToThread([chan, msg] {
+                    chan->addMessage(msg);
+                });
```
Comment (pajlada) [inline src/Application.cpp:536]: Let's make the message inside the thread like we do above

File: src/providers/twitch/PubSubManager.cpp [hunk 4]
```diff
@@ -1169,6 +1213,38 @@ void PubSub::handleMessageResponse(const PubSubMessageMessage &message)
...
+            break;
+
+            case PubSubLowTrustUsersMessage::Type::TreatmentUpdate: {
+                this->signals_.moderation.suspiciousTreatmentUpdated.invoke(
+                    innerMessage);
+            }
+            break;
+
+            case PubSubLowTrustUsersMessage::Type::INVALID: {
+                qCDebug(chatterinoPubSub)
```
Comment (pajlada) [inline src/providers/twitch/PubSubManager.cpp:1241]: Let's upgrade this to a warning

File: src/providers/twitch/PubSubManager.hpp [hunk 5]
```diff
@@ -126,12 +127,50 @@ class PubSub
...
+     * are caught by Automod.
+     *
+     * PubSub topic: automod-queue.{channelID}.{currentUserID}
+     */
     void listenToAutomod(const QString &channelID);
 
+    /**
+     * Listen to Low Trust events in the given channel.
+     * This topic is only relevant for moderators.
+     *
```
Comment (pajlada) [inline src/providers/twitch/PubSubManager.hpp:163]: Could you fill in a short blurb about what the topic is for?

File: src/providers/twitch/TwitchChannel.cpp [hunk 6]
```diff
@@ -1253,7 +1253,11 @@ void TwitchChannel::refreshPubSub()
     getApp()->twitch->pubsub->setAccount(currentAccount);
 
     getApp()->twitch->pubsub->listenToChannelModerationActions(roomId);
-    getApp()->twitch->pubsub->listenToAutomod(roomId);
+    if (this->hasModRights())
+    {
+        getApp()->twitch->pubsub->listenToAutomod(roomId);
+        getApp()->twitch->pubsub->listenToLowTrustUsers(roomId);
+    }
```
Comment (pajlada) [inline src/providers/twitch/TwitchChannel.cpp:1260]: Could you make sure to try the low trust user topic out again after this change?
I've tested it & it seems to work, but I haven't given it super much testing

File: src/providers/twitch/TwitchIrcServer.cpp [hunk 7]
```diff
@@ -45,11 +43,10 @@ TwitchIrcServer::TwitchIrcServer()
     , liveChannel(new Channel("/live", Channel::Type::TwitchLive))
     , automodChannel(new Channel("/automod", Channel::Type::TwitchAutomod))
     , watchingChannel(Channel::getEmpty(), Channel::Type::TwitchWatching)
+    , pubsub(new PubSub(TWITCH_PUBSUB_URL))
```
Comment (pajlada) [inline src/providers/twitch/TwitchIrcServer.cpp:46]: This should hopefully not change anything, but if something breaks it might be because of this

**Review by pajlada — COMMENTED** [review]
Works as expected, some small nitpicks (& the regular that I'd like you to look at the changes I've pushed)

**Comment by iProdigy** [comment]
Everything still works as expected!

image

image

**Review by pajlada — APPROVED** [review]

**Pull request merged**